c2rust-analyze/src/main.rs

```diff
@@ -449,13 +449,65 @@ fn run(tcx: TyCtxt) {
             assert_eq!(local, l);
         }
 
+        // Find all calls to `malloc`, `calloc`, `realloc`, and `free`
+        // and track their destination locals as being libc::c_void
+        // pointers to special-case downstream casts
+        for bb_data in mir.basic_blocks().iter() {
+            if let Some(term) = &bb_data.terminator {
+                let term: &Terminator = term;
+                if let TerminatorKind::Call {
+                    ref func,
+                    ref args,
+                    destination,
+                    target: _,
+                    ..
+                } = term.kind
+                {
+                    let func_ty = func.ty(&mir.local_decls, tcx);
+                    use crate::util::Callee::*;
+
+                    let assert_libc_cvoid = |p: &Place<'tcx>| {
+                        let deref_ty = p
+                            .ty(acx.local_decls, acx.tcx())
+                            .ty
+                            .builtin_deref(true)
+                            .unwrap()
+                            .ty
+                            .kind();
+                        assert_matches!(deref_ty, TyKind::Adt(adt, _) => {
+                            assert_eq!(tcx.def_path(adt.did()).data[0].to_string(), "ffi");
+                            assert_eq!(tcx.item_name(adt.did()).as_str(), "c_void");
+                        });
+                    };
+
+                    match util::ty_callee(tcx, func_ty) {
+                        Some(Malloc | Calloc) => {
+                            assert_libc_cvoid(&destination);
+                            acx.c_void_ptrs.insert(destination);
+                        }
+                        Some(Realloc) => {
+                            assert_libc_cvoid(&destination);
+                            acx.c_void_ptrs.insert(destination);
+                            acx.c_void_ptrs.insert(args[0].place().unwrap());
```

Comment:
Do you need an `assert_libc_cvoid(&args[0].place().unwrap())`, too, like for the other ones?

Reply:
Can you extract an `add_c_void_ptr` function that does both `assert_libc_cvoid` and `.insert`?

Reply:
added in 9f72789c